### ISSUE TITLE
model_registry schema design

### ISSUE THREAD

**Issue by StephenODea54** [wetherc/dstk#8]
@wetherc 
ref #2 

Wanted to start planning out the schema for the `model_registry` db. Below is what I have so far. Let me know if these are along the lines of what you had in mind!

![image](https://github.com/wetherc/dstk/assets/77359138/43fb63eb-68e6-4ca5-9213-5ea026a90a30)

- `registry_models`
  - `id`: PK
  - `name`: name for the model
  - `library`: Thinking something like "Tensorflow" or "Scikit-Learn".
  - `type`: Thinking something like "Linear Regression" or "Random Forest"
    - If we implement `library` and `type`, are these enum types or another table entirely?
  - `is_archived`: boolean indicator for if the model is archived or not.
  - `model_version_id`: FK reference to the `registry_model_versions` table.
  - `storage_provider_id`: FK reference to the `registry_storage_providers` table.
- `registry_storage_providers`
  - `id`: PK
  - `provider`: Thinking AWS, Google, Azure, etc.
  - `type`: Thinking S3, whatever GCP's storage solutions are, etc.
  - `url`: url where the storage exists
- `registry_model_versions`
  - `id`: PK
  - `date_created`: date when the model version was released
  - `model_transaction_id`: FK reference to the `registry_transactions` table
  - `parameters`: Thinking parameters for a model. Could be things like `n_estimators`, `min_samples_split`, etc.
  - `data`: Something to do with the data. Wouldn't be a giant json of the actual data, but maybe something that helps distinguish the data used between versions.
  - `dependencies`: Dependencies and their versions used in the buildup of the model (e.g. `pandas v1.x.x`)
- `registry_transactions`
  - `id`: PK
  - `transaction_date`: date for which the transaction took place.
  - `transaction_type` (CRUD)
  - `log`: The actual auditable log.

**Comment by wetherc** [comment]
I like where you're going with this! A few thoughts:

In the general case, we might want to use two separate ID columns: one for internal database use (auto-incrementing integer), and one for external use (a UUIDv4). Using UUIDs as primary keys can have some performance overhead, and using integers as externalized keys can also [cause some hiccups if you're careless](https://secure.phabricator.com/book/phabflavor/article/things_you_should_do_now/#start-ids-at-a-gigantic-number).

- For `registry_model`:
  - There's a 1:many relationship between `registry_model` and `registry_model_versions`; the model version should have the foreign key to the model, not the other way around.
  - not sure how I feel about `type`. It's not going to change how we deploy a model, and there are more nuanced-enough-to-have-their-own-name model types than we have the creativity to make an enum for
  - could we add a `description` field?
  - We'll probably want a `created_by` field to track the user that created the model
- For `registry_model_versions`:
  - let's leave off the transaction FK for now, we need to think through audit logging a bit more
  - I don't think that we necessarily benefit from `parameters`, or `data`  here. The model version is going to be the trained model as a binary blob stored in an upstream storage provider: it'll be pretty self-contained. I'd be okay with exposing a `metadata` field and letting users upload whatever valid JSON blob they want to describe model hyperparams or whatever else.
  - A `description` field is probably useful here as well
  - We'll probably want a `created_by` field to track the user that registered the model version
  - `dependencies` I'm not sure about... We'll need to think about how agnostic we are in model deployments since there's no requirement that models be trained in a Python environment
- For `registry_storage_providers`:
  - High level, I'd say let's only target S3 API-compatible storage providers for now just to limit complexity. (I.e., MinIO, S3, Ceph, etc.) For these, we'd need the following columns: `endpoint`, `region`, `bucket`, `access_key_id`, and `secret_access_key`
- let's hold off on `registry_transactions` for a minute so we can figure out how we want to track and audit stuff. I think it'll be a bit dumber than what you're proposing, but we should talk it through

**Comment by StephenODea54** [comment]
Gotcha! Yeah a lot of that makes sense. Here's an updated version based on your feedback:

![image](https://github.com/wetherc/dstk/assets/77359138/2c211f21-61ae-423f-ad91-ad9b31017bd3)

- For the same reasons for removing `Type` I have also excluded `library`.
- I'm assuming "internal database use" has everything to do with the database communicating with itself, and "external database use" has everything to do with api's communicating with the db. So I've defined all of the relations using the integer-based keys.
  - In the same light the `created_by` fields would be related to the `dtsk_user` db, so I have the `created_by` fields to be the integer values from whatever table we put in there that holds the user info.
